src/shared/auto-update-cache.ts

```diff
@@ -0,0 +1,58 @@
+import { TimeString } from '@types';
+import ms, { StringValue } from 'ms';
+import { timeoutPromise } from './timeouts';
+type Timestamp = number;
+type CacheConstructorParams<Value> = {
+  calculate: () => Promise<Value>;
+  config: AutoUpdateCacheConfig;
+};
+
+export type AutoUpdateCacheConfig = {
+  options: AutoUpdateConfigOptions;
+};
+
+export type AutoUpdateConfigOptions = {
+  useCachedValue: 'always' | { ifUnder: StringValue };
+  nextUpdate: { ifSuccess: StringValue; ifFails: StringValue };
```

Comment:
Small suggestion. A little more verbose, but in my opinion a little clearer. Also, we could make `ifFailsTryAgainIn` optional and have a default. Maybe 5 min?
```suggestion
  update: { every: TimeString; ifFailsTryAgainIn?: TimeString };
```